keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+  endpoints in the future.
+
+### Configuration
+
+A new `topology.kubernetes.io/topology-aware-routing` annotation will be
+supported on Services. This will have 2 potential values:
+
+- `Auto`: When there are a sufficient number of endpoints for the Service, the
+  EndpointSlice controller will add topology hints for each endpoint that will
+  ensure a proportional amounts are available to each zone in a cluster.
+- `Disabled`: EndpointSlice hints will not be set for this feature.
+
+When this annotation is either unspecified or not set to `Auto`, the `Disabled`
+behavior will be default. A future KEP will explore changing the default
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and `topologyKeys` are set on a Service, `topologyKeys`
+will be given precedence and this annotation will be ignored.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `TopologyHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // zone indicates the zone this endpoint should be consumed by to enable
+  // topology aware routing.
+  // +optional
+  Zone string `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+}
+```
+
+#### Future API Expansion
+This approach would allow for future API expansion that enabled specifying
+multiple zones per endpoint with weights. That level of complexity may never be
+necessary, but it will be possible. For example:
+
+```yaml
+topologyHints:
+  zones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
+updated to filter endpoints based on topology hints when the following
+conditions are true:
+
+- Kube-Proxy is able to determine the zone it is running within (likely based
+  on node labels).
+- The `topology.kubernetes.io/topology-aware-routing` annotation is set to
+  `Auto` for the Service.
+- At least one endpoint for the Service has a hint pointing to the zone
+  Kube-Proxy is running within.
+- All endpoints for the Service have zone hints.
+
+When the above conditions are true, kube-proxy will only route traffic to
+endpoints with a hint referring to the zone Kube-Proxy is running within.
```

Comment:
fallback - if all endpoints are hinted but none are for me - fall back on random?

Reply:
That's what I was trying to communicate here.  One of the conditions above is "At least one endpoint for the Service has a hint pointing to the zone Kube-Proxy is running within". I'll try to make that clearer in the KEP.

Reply:
Added a note below this sentence to clarify this.